Stratis.Bitcoin.Features.BlockStore/LoopSteps/ProcessPendingStorageStep.cs

```diff
@@ -43,132 +42,158 @@ internal ProcessPendingStorageStep(BlockStoreLoop blockStoreLoop, ILoggerFactory
-                if (blockIsInPendingStorage == false)
-                {
-                    if (context.PendingBlockPairsToStore.Count == 0)
-                        return StepResult.Next;
+                PrepareNextBlockFromPendingStorage(context);
 
-                    if (context.PendingBlockPairsToStore.Any())
-                    {
-                        this.logger.LogTrace("[FLUSH_NEXTBLOCKNOTINPENDING]", nameof(context.PendingStorageBatchSize), context.PendingStorageBatchSize);
-                        await this.PushBlocksToRepository(context);
+                var canProcessNextBlock = context.CanProcessNextBlock();
+                if (canProcessNextBlock == false)
+                    if (!context.PendingBlockPairsToStore.Any())
                         break;
-                    }
-                }
-                else
-                {
-                    context.PendingBlockPairsToStore.Push(context.PendingBlockPairToStore);
-                    context.PendingStorageBatchSize += context.PendingBlockPairToStore.Block.GetSerializedSize();
-                }
+            }
 
-                if (disposeMode)
-                {
-                    this.logger.LogTrace("{0}:{1} [FLUSH_BLOCKSTORE_DISPOSING]", nameof(context.PendingStorageBatchSize), context.PendingStorageBatchSize);
-                    await this.PushBlocksToRepository(context);
-                    break;
-                }
+            if (context.PendingBlockPairsToStore.Any())
+                await PushBlocksToRepository(context);
 
-                if (context.PendingStorageBatchSize > BlockStoreLoop.MaxPendingInsertBlockSize)
-                {
-                    this.logger.LogTrace("{0}:{1} [FLUSH_BATCHSIZE_REACHED]", nameof(context.PendingStorageBatchSize), context.PendingStorageBatchSize);
-                    await this.PushBlocksToRepository(context);
-                    break;
-                }
+            return StepResult.Stop;
+        }
 
-                if (this.BlockStoreLoop.ChainState.IsInitialBlockDownload == false)
-                {
-                    this.logger.LogTrace("[FLUSH_NOT_IN_IBD]");
-                    await this.PushBlocksToRepository(context);
-                    break;
-                }
+        /// <summary>
+        /// When the node is in IBD continuously process all the blocks in <see cref="BlockStoreLoop.PendingStorage"/> until 
+        /// a stop condition is found.
+        /// </summary>
+        private async Task<StepResult> ProcessWhenInIBD(ProcessPendingStorageContext context)
+        {
+            this.logger.LogDebug("{0}:'{1}/{2}'", nameof(context.NextChainedBlock), context.NextChainedBlock.HashBlock, context.NextChainedBlock.Height);
+
+            if (this.BlockStoreLoop.PendingStorage.Count < BlockStoreLoop.PendingStorageBatchThreshold)
+                return StepResult.Continue;
 
-                context.GetNextBlock();
+            do
```

Comment:
no need for do;while see other comment

Reply:
As per your other comment I have changed this to a `while { }`